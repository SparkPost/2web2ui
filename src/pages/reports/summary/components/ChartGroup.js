import React, { Component } from 'react';
import _ from 'lodash';
import { getDayLines, getLineChartFormatters } from 'src/helpers/chart';
import LineChart from './LineChart'; // async
import METRICS_UNIT_CONFIG from 'src/config/metrics-units';
const DEFAULT_UNIT = 'number';

function getUniqueUnits(metrics) {
  return _.uniq(metrics.map(({ unit = DEFAULT_UNIT }) => unit));
}

export default class ChartGroup extends Component {
  createDayReferenceLines() {
    const { chartData, precision } = this.props;

    return getDayLines(chartData, precision).map(({ ts }) => ({
      key: ts,
      x: ts,
      stroke: '#bbb',
      strokeWidth: 2,
    }));
  }

  render() {
    const { chartData = [], metrics, chartLoading, precision, yScale, to } = this.props;

    if (!chartData.length || !metrics) {
      return null;
    }

    const formatters = getLineChartFormatters(precision, to);
    const referenceLines = this.createDayReferenceLines();
    const charts = getUniqueUnits(metrics).map(unit => ({
      metrics: metrics.filter(metric => metric.unit === unit),
      ...METRICS_UNIT_CONFIG[unit],
    }));

    return (
      <div>
        {charts.map((chart, i) => (
          <LineChart
            key={`chart=${i}`}
            syncId="summaryChart"
            data={chartData}
            precision={precision}
            lines={chart.metrics.map(({ name, label, stroke }) => ({
              key: name,
              dataKey: name,
              name: label,
              stroke: chartLoading ? '#f8f8f8' : stroke,
            }))}
            {...formatters}
            yTickFormatter={chart.yAxisFormatter}
            yScale={yScale}
            yLabel={chart.label}
            tooltipValueFormatter={chart.yAxisFormatter}
            referenceLines={referenceLines}
            showXAxis={i === charts.length - 1}
          />
        ))}
      </div>
    );
  }
}
