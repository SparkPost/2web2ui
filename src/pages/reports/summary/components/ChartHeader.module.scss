@import '~@sparkpost/matchbox/src/styles/config.scss';

.ChartHeader {
  margin-bottom: spacing(large);
}

.TimeMode {
  display: inline-block;
  font-size: rem(13);
  line-height: rem(15);
  color: color(gray, 4);
  padding: spacing(small) 0 0;
}

.Controls {
  margin-top: spacing();
  margin-left: spacing(larger);
  text-align: left;

  & * {
    text-align: left;
  }

  @media screen and (min-width: breakpoint(small)) {
    margin-top: 0;
    margin-left: 0;
    text-align: right;
  }
}

.ButtonSpacer {
  margin-left: spacing(small);
}
