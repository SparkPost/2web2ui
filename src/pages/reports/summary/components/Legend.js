import React from 'react';
import { connect } from 'react-redux';
import getConfig from 'src/helpers/getConfig';
import styles from './Legend.module.scss';

function renderMetric(metric, uniqueLabel) {
  return (
    <div className={styles.Metric} key={metric.name}>
      {/* eslint-disable-next-line */}
      <span className={styles.Color} style={{ backgroundColor: metric.stroke }} />
      {metric.isUniquePerTimePeriod && uniqueLabel
        ? `${metric.label} ${uniqueLabel}`
        : metric.label}
    </div>
  );
}

export const Legend = props => {
  const { metrics, reportOptions } = props;
  const { precision } = reportOptions;

  const metricsRollupPrecisionMap = getConfig('metricsRollupPrecisionMap');
  const precisionObj = metricsRollupPrecisionMap.find(p => p.value === precision);
  return (
    <div className={styles.Legend}>
      {metrics &&
        metrics.map(metric => renderMetric(metric, precisionObj && precisionObj.uniqueLabel))}
    </div>
  );
};

const mapStateToProps = state => ({
  reportOptions: state.reportOptions,
});

export default connect(mapStateToProps)(Legend);
