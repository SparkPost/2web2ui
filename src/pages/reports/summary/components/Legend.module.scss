@import '~@sparkpost/matchbox/src/styles/config.scss';

.Legend {
  margin-top: rem(3);
  margin-left: spacing(larger);
}

.Metric {
  display: inline-block;
  vertical-align: middle;
  margin-right: spacing(large);
  font-weight: 400;
}

.Color {
  display: inline-block;
  vertical-align: middle;
  margin-right: rem(9);
  margin-top: -2px;
  height: 14px;
  width: 14px;
  border-radius: border-radius();
}
