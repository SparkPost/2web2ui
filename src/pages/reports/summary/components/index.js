export { default as ChartGroup } from './ChartGroup';
export { default as ChartHeader } from './ChartHeader';
export { default as Legend } from './Legend';
export { default as Table } from './Table';
export { default as MetricsModal } from './MetricsModal';
