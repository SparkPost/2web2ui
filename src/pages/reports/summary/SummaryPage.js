import React, { Component } from 'react';
import { connect } from 'react-redux';
import classnames from 'classnames';
import { refreshSummaryReport } from 'src/actions/summaryChart';
import { refreshReportOptions } from 'src/actions/reportOptions';
import { Page, Panel } from 'src/components/matchbox';
import { Loading } from 'src/components';
import ReportOptions from '../components/ReportOptions';
import { Table, MetricsModal, ChartGroup, ChartHeader } from './components';
import { selectSummaryChartSearchOptions } from 'src/selectors/reportSearchOptions';
import { selectReportsEnhancementsEnabled } from 'src/selectors/customReports';
import styles from './SummaryPage.module.scss';
import { GUIDE_IDS } from 'src/constants';

export class SummaryReportPage extends Component {
  state = {
    metricsModal: false,
    eventTime: 'real',
    scale: 'linear',
  };

  componentDidUpdate(prevProps) {
    if (prevProps.reportOptions !== this.props.reportOptions) {
      this.props.refreshSummaryReport(this.props.reportOptions);
    }

    if (prevProps.location?.state?.triggerGuide) {
      if (window.Appcues) {
        window.Appcues.show(GUIDE_IDS.EXPLORE_ANALYTICS);
      }
    }
  }

  renderLoading() {
    const { chart } = this.props;
    if (chart.chartLoading) {
      return (
        <div className={styles.Loading}>
          <Loading />
        </div>
      );
    }
  }

  handleMetricsApply = selectedMetrics => {
    this.setState({ metricsModal: false });
    this.props.refreshReportOptions({ metrics: selectedMetrics });
  };

  handleMetricsModal = () => {
    this.setState({ metricsModal: !this.state.metricsModal });
  };

  handleTimeClick = time => {
    this.setState({ eventTime: time });
  };

  handleScaleClick = scale => {
    this.setState({ scale });
  };

  render() {
    const { chart, summarySearchOptions = {}, enhancementsEnabled } = this.props;
    const { to } = summarySearchOptions;
    const { scale, eventTime, metricsModal } = this.state;

    return (
      <Page title="Summary Report">
        <ReportOptions
          reportLoading={chart.chartLoading}
          searchOptions={summarySearchOptions}
          customReportsEnabled={enhancementsEnabled}
        />
        <div data-id="summary-chart">
          <Panel.LEGACY>
            <Panel.LEGACY.Section
              className={classnames(styles.ChartSection, chart.chartLoading && styles.pending)}
            >
              <ChartHeader
                selectedMetrics={chart.metrics}
                selectedTime={eventTime}
                selectedScale={scale}
                onScaleClick={this.handleScaleClick}
                onTimeClick={this.handleTimeClick}
                onMetricsToggle={this.handleMetricsModal}
              />
              <ChartGroup {...chart} to={to} yScale={scale} />
            </Panel.LEGACY.Section>

            {this.renderLoading()}
          </Panel.LEGACY>
        </div>
        <div data-id="summary-table">
          <Table />
        </div>

        <MetricsModal
          selectedMetrics={chart.metrics}
          open={metricsModal}
          onCancel={this.handleMetricsModal}
          onSubmit={this.handleMetricsApply}
        />
      </Page>
    );
  }
}

const mapStateToProps = state => ({
  chart: state.summaryChart,
  reportOptions: state.reportOptions,
  summarySearchOptions: selectSummaryChartSearchOptions(state),
  enhancementsEnabled: selectReportsEnhancementsEnabled(state),
});

const mapDispatchToProps = {
  refreshReportOptions,
  refreshSummaryReport,
};

export default connect(mapStateToProps, mapDispatchToProps)(SummaryReportPage);
