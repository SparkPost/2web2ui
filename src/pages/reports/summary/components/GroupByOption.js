import React, { Component } from 'react';
import styles from './Table.module.scss';
import { Grid, Checkbox, Select } from 'src/components/matchbox';
import _ from 'lodash';
import { GROUP_CONFIG } from './tableConfig';

export class GroupByOption extends Component {
  state = {
    topDomainsOnly: true,
  };

  handleGroupChange = e => {
    this.props._getTableData({ groupBy: e.target.value });
  };

  handleDomainsCheckboxChange = () => {
    const topDomainsOnly = !this.state.topDomainsOnly;
    this.setState({ topDomainsOnly });
    const groupBy = topDomainsOnly ? 'watched-domain' : 'domain';
    this.props._getTableData({ groupBy });
  };

  getSelectOptions = () => {
    const { hasSubaccounts } = this.props;
    const { topDomainsOnly } = this.state;

    const filteredOptionsKeys = _.reduce(
      GROUP_CONFIG,
      (accumulator, value, key) => {
        if (
          (key === 'subaccount' && !hasSubaccounts) ||
          (key === 'domain' && topDomainsOnly) ||
          (key === 'watched-domain' && !topDomainsOnly)
        ) {
          return accumulator; // ignore
        }
        accumulator.push(key);
        return accumulator;
      },
      [],
    );

    const options = filteredOptionsKeys.map(key => ({
      value: key,
      label: GROUP_CONFIG[key].label,
    }));

    return options;
  };

  renderDomainsCheckbox() {
    const { tableLoading, groupBy } = this.props;
    const { topDomainsOnly } = this.state;

    //Only show 'Top Domains Only' checkbox when on the recipient domains grouping
    if (groupBy !== 'watched-domain' && groupBy !== 'domain') {
      return null;
    }
    return (
      <div className={styles.TopDomainsCheckbox}>
        <Checkbox
          id="watchedDomains"
          label="Top Domains Only"
          checked={topDomainsOnly}
          onChange={this.handleDomainsCheckboxChange}
          disabled={tableLoading}
        />
      </div>
    );
  }

  render() {
    const { groupBy, tableLoading } = this.props;
    return (
      <Grid>
        <Grid.Column xs={12} md={5} lg={4}>
          <Select
            label="Group By"
            options={this.getSelectOptions()}
            value={groupBy}
            disabled={tableLoading}
            onChange={this.handleGroupChange}
          />
        </Grid.Column>
        <Grid.Column xs={12} md={4} mdOffset={3} lg={3} lgOffset={5}>
          {this.renderDomainsCheckbox()}
        </Grid.Column>
      </Grid>
    );
  }
}

export default GroupByOption;
