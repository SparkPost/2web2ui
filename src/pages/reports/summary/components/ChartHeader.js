import React, { Component } from 'react';
import { Grid, Button } from 'src/components/matchbox';
import Legend from './Legend';

import styles from './ChartHeader.module.scss';

export default class ChartHeader extends Component {
  renderScaleButton(scale, label) {
    const { onScaleClick, selectedScale } = this.props;
    return (
      <Button size="small" primary={scale === selectedScale} onClick={() => onScaleClick(scale)}>
        {label}
      </Button>
    );
  }

  renderTimeMode() {
    const { selectedTime, onTimeClick } = this.props;
    return (
      <Button.Group className={styles.ButtonSpacer}>
        <Button primary={selectedTime === 'real'} onClick={() => onTimeClick('real')} size="small">
          Real Time
        </Button>
        <Button
          primary={selectedTime === 'injection'}
          onClick={() => onTimeClick('injection')}
          size="small"
        >
          Injection Time
        </Button>
      </Button.Group>
    );
  }

  render() {
    const { selectedMetrics, onMetricsToggle } = this.props;

    return (
      <Grid className={styles.ChartHeader}>
        <Grid.Column xs={12} md={9}>
          <Legend metrics={selectedMetrics} />
        </Grid.Column>
        <Grid.Column xs={12} md={3}>
          <div className={styles.Controls}>
            <Button size="small" onClick={onMetricsToggle}>
              Select Metrics
            </Button>
            {/* {this.renderTimeMode()} */}
            {/*<Button.Group className={styles.ButtonSpacer}>*/}
            {/*{this.renderScaleButton('linear', 'Linear')}*/}
            {/*{this.renderScaleButton('log', 'Log')}*/}
            {/*{this.renderScaleButton('pow', 'Pow')}*/}
            {/*</Button.Group>*/}
          </div>
        </Grid.Column>
      </Grid>
    );
  }
}
