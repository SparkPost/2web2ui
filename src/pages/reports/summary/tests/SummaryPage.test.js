import React from 'react';
import { shallow } from 'enzyme';
import { SummaryReportPage } from '../SummaryPage';
import * as reportHelpers from 'src/helpers/reports';
import { MetricsModal, ChartHeader } from '../components';
import { GUIDE_IDS } from 'src/constants';
jest.mock('src/helpers/reports');

describe('Page: SummaryPage', () => {
  let wrapper;
  let testProps;
  const mockshow = jest.fn();

  beforeEach(() => {
    reportHelpers.getFilterSearchOptions = jest.fn(() => ({}));
    reportHelpers.parseSearch = jest.fn(() => ({ options: {} }));
    testProps = {
      reportOptions: [],
      chart: {
        metrics: [1, 2, 3],
      },
      refreshSummaryReport: jest.fn(),
      refreshReportOptions: jest.fn(),
      addFilters: jest.fn(),
      location: { state: { triggerGuide: true } },
    };
    wrapper = shallow(<SummaryReportPage {...testProps} />);
    window.Appcues = {};
    window.Appcues = { show: mockshow };
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should refresh when report options reference changes', () => {
    const newReportOptions = {};
    expect(testProps.refreshSummaryReport).not.toHaveBeenCalled();
    wrapper.setProps({ reportOptions: newReportOptions });
    expect(testProps.refreshSummaryReport).toHaveBeenCalledWith(newReportOptions);
  });

  it('should trigger guide when location state has triggerGuide set to true', () => {
    const location = { state: { triggerGuide: true } };
    expect(mockshow).not.toHaveBeenCalled();
    wrapper.setProps({ location: location });
    expect(mockshow).toHaveBeenCalledWith(GUIDE_IDS.EXPLORE_ANALYTICS);
  });

  it('should render when loading', () => {
    testProps.chart.chartLoading = true;
    wrapper = shallow(<SummaryReportPage {...testProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle metrics modal chart header', () => {
    const header = wrapper.find(ChartHeader);
    expect(wrapper.state('metricsModal')).toEqual(false);
    header.simulate('metricsToggle');
    expect(wrapper.state('metricsModal')).toEqual(true);
    header.simulate('metricsToggle');
    expect(wrapper.state('metricsModal')).toEqual(false);
  });

  it('should toggle modals via metrics modal cancel', () => {
    const modal = wrapper.find(MetricsModal);
    expect(wrapper.state('metricsModal')).toEqual(false);
    modal.simulate('cancel');
    expect(wrapper.state('metricsModal')).toEqual(true);
    modal.simulate('cancel');
    expect(wrapper.state('metricsModal')).toEqual(false);
  });

  it('should apply metrics via metrics modal submit', () => {
    wrapper.setState({ metricsModal: true });
    const modal = wrapper.find(MetricsModal);
    const metrics = {};
    modal.simulate('submit', metrics);
    expect(testProps.refreshReportOptions).toHaveBeenCalledWith({ metrics });
    expect(wrapper.state('metricsModal')).toEqual(false);
  });

  it('should handle scale changes', () => {
    const header = wrapper.find(ChartHeader);
    header.simulate('scaleClick', 'test-scale');
    expect(wrapper.state('scale')).toEqual('test-scale');
  });

  it('should handle time changes', () => {
    const header = wrapper.find(ChartHeader);
    header.simulate('timeClick', 'test-time');
    expect(wrapper.state('eventTime')).toEqual('test-time');
  });
});
