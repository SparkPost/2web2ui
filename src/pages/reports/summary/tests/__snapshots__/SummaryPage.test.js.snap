// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: SummaryPage should render correctly 1`] = `
<Page
  title="Summary Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <div
    data-id="summary-chart"
  >
    <Panel.LEGACY>
      <Panel.LEGACY.Section
        className="ChartSection"
      >
        <ChartHeader
          onMetricsToggle={[Function]}
          onScaleClick={[Function]}
          onTimeClick={[Function]}
          selectedMetrics={
            Array [
              1,
              2,
              3,
            ]
          }
          selectedScale="linear"
          selectedTime="real"
        />
        <ChartGroup
          metrics={
            Array [
              1,
              2,
              3,
            ]
          }
          yScale="linear"
        />
      </Panel.LEGACY.Section>
    </Panel.LEGACY>
  </div>
  <div
    data-id="summary-table"
  >
    <Connect(Table) />
  </div>
  <MetricsModal
    onCancel={[Function]}
    onSubmit={[Function]}
    open={false}
    selectedMetrics={
      Array [
        1,
        2,
        3,
      ]
    }
  />
</Page>
`;

exports[`Page: SummaryPage should render when loading 1`] = `
<Page
  title="Summary Report"
>
  <withRouter(Connect(ReportOptions))
    reportLoading={true}
    searchOptions={Object {}}
  />
  <div
    data-id="summary-chart"
  >
    <Panel.LEGACY>
      <Panel.LEGACY.Section
        className="ChartSection pending"
      >
        <ChartHeader
          onMetricsToggle={[Function]}
          onScaleClick={[Function]}
          onTimeClick={[Function]}
          selectedMetrics={
            Array [
              1,
              2,
              3,
            ]
          }
          selectedScale="linear"
          selectedTime="real"
        />
        <ChartGroup
          chartLoading={true}
          metrics={
            Array [
              1,
              2,
              3,
            ]
          }
          yScale="linear"
        />
      </Panel.LEGACY.Section>
      <div
        className="Loading"
      >
        <Loading />
      </div>
    </Panel.LEGACY>
  </div>
  <div
    data-id="summary-table"
  >
    <Connect(Table) />
  </div>
  <MetricsModal
    onCancel={[Function]}
    onSubmit={[Function]}
    open={false}
    selectedMetrics={
      Array [
        1,
        2,
        3,
      ]
    }
  />
</Page>
`;
